jigna/web_server.py

```diff
@@ -115,6 +119,126 @@ def _handlers_default(self):
     def __bridge_default(self):
         return WebBridge()
 
+    def _get_attribute_values(self, obj, attribute_names):
+        """ Get the values of all 'public' attributes on an object.
+
+        Return a list of strings.
+
+        """
+
+        return [self._marshal(self._get_attribute_default(obj, name))
+                for name in attribute_names]
+
+    def _get_attribute_default(self, obj, name):
+        value = getattr(obj, name)
+        if isinstance(value, list):
+            value = []
+        elif isinstance(value, dict):
+            value = {}
+        elif hasattr(value, '__dict__'):
+            pass
+        elif value is None:
+            pass
+        else:
+            value = type(value)()
+        return value
+
+    def _get_dict_info(self, obj):
+        """ Get a description of a dict. """
+        if self.async:
+            values = self._get_list_info(list(obj.values()))
+            return dict(keys=list(obj.keys()), values=values)
+        else:
+            return super(WebServer, self)._get_dict_info(obj)
+
+    def _get_instance_info(self, obj):
+        """ Get a description of an instance. """
+
+        info = super(WebServer, self)._get_instance_info(obj)
+
+        if self.async:
+            # If this is a new type, also send the attribute_values.
+            if 'attribute_names' in info:
+                attribute_values = self._get_attribute_values(
+                    obj, info['attribute_names']
+                )
+                info['attribute_values'] = attribute_values
+
+        return info
+
+    def _get_list_info(self, obj):
+        """ Get a description of a list. """
+        if self.async:
+            data = self._marshal_all(obj)
+            new_types = {}
+            for x in data:
+                if x['type'] == 'instance' and len(x['info']) > 1:
+                    new_types[x['info']['type_name']] = x['info']
+
+            return dict(length=len(obj), data=data, new_types=new_types)
+        else:
+            return super(WebServer, self)._get_list_info(obj)
+
+    def _send_object_changed_event(self, obj, trait_name, old, new):
+        """ Send an object changed event. """
+
+        if not self.async:
+            return super(WebServer, self)._send_object_changed_event(
+                obj, trait_name, old, new
+            )
+
+        # Otherwise send all the async related info.
+
+        if trait_name.startswith('_'):
+            return
```

Comment:
This does not handle complex slicing like `list[::2] = [1,2,3]`  At least document this.